### PR TITLE
Initial seeding from MP sandbox

### PR DESCRIPTION
**Pull request by hutchig**
Signed-off-by: Gordon Hutchison <Gordon.Hutchison@gmail.com>

### CODE REVIEW

**Comment by kenfinnigan** [comment]
Did you mean for this to be in a directory structure of /proposals/event-sourcing? I think we can just have it at the root.

**Comment by hutchig** [comment]
@kenfinnigan I had mirrored the structure of the microprofile-sandbox, but I think you are right so have moved the files up towards the root.

**Comment by kenfinnigan** [comment]
Is the idea that what's in /proposal becomes a possible spec doc?

I think initially it's probably easier to have all that information/docs in the main repo README, no?

The main piece of work here is developing a set of APIs and then implementing them. Whatever might constitute a "spec doc" would be the last piece

**Comment by hutchig** [comment]
I consider the README, proposal and spec docs to be different documents.

The README's job is to be the readable webpage/landing page for the repo - an attractive open door.

The proposal is a document making the case to exit the sandbox and become a proper MicroProfile project, it is not a draft specification document. For example, topics like value, approach and possible future directions and others that are useful in a proposal do not belong in a specification, let alone a 1.0 specification.

The specification would be developed only based on our findings and subsequent work as an accepted
project group and ideally be created with an open process of issue/PRs/hangouts with minutes (if desired) and so on. I don't think this repo is a place to develop a MP Specification - that work belongs in github.com/eclipse/microprofile-event-sourcing if we get an accepted proposal.

**Comment by hutchig** [comment]
@kenfinnigan  I am happy to start with a blank page for the draft proposal doc
developed 100% as a group effort but at this stage I am just dumping ideas and
trying to raise interest.

I believe we need some readable text in order to share the idea of the spec area
in order to help to have an sigpost/open door to whomever wants to get involved. 
I am hoping (if we get to it) to raise the topic area in today's community hangout
and needed a doc at a URL for interested people to be able to read.

After it gets some time in a mp community hangout - I suggest we organize an initial
event sourcing group hangout to discuss how we(however that is) work.
I would love the proposal doc to be 'mined out' using PRs and paragraphs
turned into issues that in-turn become code/text PRs.

I note that the topic of an event sourcing PoC and proposal workgroup
is now after your agenda item to block people creating proposals
without first having implementations.  As I have said before I have no
intention of submitting any proposal before some PoC work.

One of the reasons I choose 'proposal.asciidoc' and the sandbox was exactly because it
generates lots of attention. I once watched as Mark Struberg got
lots of energy from the community because he used the word
specification 'too early' and a slightly presumptious package name in some strawman work
he was doing. Lots of people popped up and said
he had no right to do it - but it kickstarted the effort which, in the end,
was completed to great effect.

**Comment by kenfinnigan** [comment]
In terms of dumping ideas, the best way to do that is to create issues.

You mention that the proposal doc is making a case for exiting the sandbox. We're not "in" the sandbox at present. This is meant to be for an implementation, once we have an implementation we can then revisit creating of a sandbox proposal based on what was learnt in the implementation

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,364 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
```
Comment (kenfinnigan) [inline pom.xml]: Please remove the copyright notice, we try not to add those to pom files as they're not necessary

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.smallrye</groupId>
```
Comment (kenfinnigan) [inline pom.xml:6]: Need to inherit from `smallrye-parent`, then most of the other plugin setup can be removed

File: pom.xml [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    <artifactId>smallrye-event-sourcing</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
+    
+    <name>MicroProfile Event Sourcing Proof Of Concept</name>
```
Comment (kenfinnigan) [inline pom.xml]: I would suggest calling it "SmallRye Event Sourcing".

What we're doing might end up being a PoC for MP, but that's not what this repo is. It's an actual implementation of Event Sourcing in SmallRye

File: pom.xml [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+     <name>SmallRye</name>
+     <url>https://wwww.smallrye.io</url>
+   </organization>
+
+    <issueManagement>
+        <system>GitHub</system>
+        <url>https://github.com/smallrye/smallrye-event-sourcing/issues</url>
+    </issueManagement>
+
+    <distributionManagement>
```
Comment (kenfinnigan) [inline pom.xml]: This will come from the parent

File: pom.xml [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+       -->
+    </modules>
+
+    <dependencyManagement>
+        <dependencies>
+        </dependencies>
+    </dependencyManagement>
+
+    <build>
+        <pluginManagement>
```
Comment (kenfinnigan) [inline pom.xml]: All these plugin definitions can be removed when using the parent

File: pom.xml [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                            </attributes>
+                        </configuration>
+                    </execution>
+                </executions>
+                
+            </plugin>
+
+        </plugins>
+    </build>
+    <profiles>
```
Comment (kenfinnigan) [inline pom.xml]: Profile can be removed when using the parent

File: pom.xml [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.smallrye</groupId>
+    <artifactId>smallrye-event-sourcing</artifactId>
+    <version>1.0.0-SNAPSHOT</version>
+    <packaging>pom</packaging>
```
Comment (kenfinnigan) [inline pom.xml]: I think initially the packaging should be `jar` and everything under /proposal is removed.

We can separate API classes by package name initially, and then split out into separate API and implementation artifacts later.

**Review by kenfinnigan — CHANGES_REQUESTED** [review]

**Comment by hutchig** [comment]
@kenfinnigan  @cescoffier I have moved the ideas/discussion material into the readme and made the draft proposal document a blank slate as suggested. Lets have a discussion about how to proceed. Shall we schedule a videocon sometime?  Perhaps after today, as the 'rules' for proposals
might change from the hangout and I have put this in the adgenda too. 

What makes a good proposal?

I was pending Gunnar's opinion about where/how to work a little, but imagine you might have spoken to him in the meantime. Let me know any other changes you want in the PR.

**Comment by kenfinnigan** [comment]
@hutchig this isn't an MP proposal, so we don't need to do anything in the MP community.

What we're doing here is implementing an idea that might one day be part of MP.

I don't think there's need for any calls, we can do everything async via GH issues

File: pom.xml (hunk 1 above)
Comment (hutchig) [reply pom.xml]: Will do.

File: pom.xml (hunk 2 above)
Comment (hutchig) [reply pom.xml:6]: Will do.

File: pom.xml (hunk 7 above)
Comment (hutchig) [reply pom.xml]: OK will do.

File: pom.xml (hunk 4 above)
Comment (hutchig) [reply pom.xml]: Ok will do.

File: pom.xml (hunk 6 above)
Comment (hutchig) [reply pom.xml]: OK will do. Currently this is a side-project activity for me, so I will get to it later on today :-) .

Comment (kenfinnigan) [reply pom.xml]: Not a problem, there's no rush

File: pom.xml (hunk 3 above)
Comment (hutchig) [reply pom.xml]: Will do.

File: pom.xml (hunk 5 above)
Comment (hutchig) [reply pom.xml]: Ok

**Comment by hutchig** [comment]
@kenfinnigan Could it not be good audience for seeing if anyone (especially other vendors than IBM-Redhat and users/customers) might be interested in being involved?

Either way, once the idea comes up once in the hangout (today hopefully) I will be happy to follow your lead from then on, while keeping an IBM green-light in the corner of my eye of course.

I would rather make SOMETHING 1.0 happen than be too opinionated about the mechanics/control
of exactly what/how it is done, #validatedlearning :-)

**Comment by kenfinnigan** [comment]
I'm not saying don't ask for those in the MP community to help contribute, just that right now it's not an MP spec to be discussed/agreed.

We're working on an implementation to see where it goes. It may even lead to us saying there really isn't anything useful above what's already available. We won't know until work is begun

**Comment by gunnarmorling** [comment]
Hey @hutchig, first of all, thanks for interest in CDC and your efforts for exploring how this could be brought together with mP, thinking about standardization etc. That's much appreciated!

That said, it's still a bit blurry to me how all these different ideas and concepts brought up in this proposal fit together and what it actually is that would be built here. I.e. I could need some guidance on what you'd envision to build :)

On a conceptual level, I feel there's a disconnect between the notions of "event sourcing" and "change data capture". In the former, you'll typically work with higher level "business" events, which describe an event happening in the reality (e.g. "customer placed a purchase order") which may lead to the alteration of state of multiple entities (e.g. a new order gets created, the customer's credit limit gets reduced etc.). Typically, those events also might convey some kind of intent and contextual information such as a use case identifier, application user etc. In CDC on the other hand, as I see it, events are more technically oriented and fine-grained, e.g. the UPDATE of a customer, or the INSERT of an order. So one might derive CDC-style events from the ES-kind events, but it will be hard to do the opposite due to the lack of contextual information when receiving CDC events from the logs of a database. Both concepts, ES and CDC, have their place and use cases, but I'm not sure to what extent it's possible or sensible to try and implement a true ES-style architecture using CDC.

I'm bringing this point up so to better understand what you are after here. So, perhaps it's the best to illustrate the ideas in some strawman code examples here in the proposal? I.e. how would code using that suggested API look like, and what would it achieve? Details are not too important at this point, but some basic example  might help folks here to grasp the idea.

Another idea you brought up recently (and I'm not sure whether it's meant to be explored here as well or somewhere else) was to expose CDC connectors such as Debezium's as "sources" for mP Reactive Messaging. That's definitely something I could see as being very useful and it's also less blurry to me. What you described in the "Library of shared Connectors to plug-into MicroProfile Reactive Messaging as Stream sources/sinks" thread makes lots of sense to me and that'd be a great thing to explore. The biggest hurdle is management of the internal state of these CDC stream sources, as they need to persist the position in the source log they've read last in order to continue from there after an application restart.

If you are interested, I can dig out some related exploration we did a while ago based on RxJava and Debezium's embedded engine (it runs connectors as a library within a Java application, invoking a configured callback for each emitted event, i.e. there's no inherent need for something like Apache Kafka, although one common use case of this is to submit CDC events to brokers other than Kafka with a bit of glue gode, e.g.to Kinesis or Pulsar).

**Comment by hutchig** [comment]
Hi @gunnarmorling you are absolutely right the readme text in this PR contains an absolute dump
of lots of ideas (I have shared everything I have thought of (much of it second hand of course).

As I see it the primary first goal of this project is to get a working
https://download.eclipse.org/microprofile/microprofile-reactive-messaging-1.0-RC1/apidocs/org/eclipse/microprofile/reactive/messaging/spi/IncomingConnectorFactory.html
that is fed from Debezium sourced CDC events.
This can be used to for an inital PoC (milestone1.0?) where such CDC events are feeding into application code and processed using reactive streams methods. 
We might learn a lot about what might belong in a subclass of https://download.eclipse.org/microprofile/microprofile-reactive-messaging-1.0-RC1/apidocs/org/eclipse/microprofile/reactive/messaging/Message.html that is more aware of change events 
or whether we should do this in processing the payload
or via type-converters which we want in reactive message 1.next anyway. 


milestone2.0: We could create a OutgoingConnectorFactory that reproduces
the wire format the Debezium's Kafka Connector can consume on
the other end. That would allow 'applications' to create reactive messages
that feed Debeziums remote state changing code.

milestone3.0: After that we can explore what might happen on a
CDC event receiving IncomingConnectorFactory that can be used
to feed a CDC event via reactive messaging consumption into a
application. 

milestone4.0
It might be interesting for us to explore how 'nested' message envelopes
work in reactive messaging     {Debezium{Kafka{Payload}} so that we can
swap Kafka for something else.

The rest of the stuff is just bluesky thinking.

What if I proceed by creating some issues that might be in the scope
of milestone1.0 (and deleting the parts of the readme document that relate
to each issue in turn as the issue is created). 
When I start doing this I will move the readme file into something like
ideasDump and have a more sensible readme.

**Comment by hutchig** [comment]
I agree that you could not generally build a ES system out of CDC like events and that the scale, intent and context is generally different. In some senses it is 'cheeky' appropriating the name/brand of a architectural approach called 'event sourcing' - I recognise that.  I think we are really building some slightly nicer plumbing. I was sort of after a better name at the bottom of
https://groups.google.com/d/msg/microprofile/F0ehhd1MFMc/UPFcUOjGBwAJ
and I don't have one. It seems to attract the right sort of interest from the right
sort of people for now, even if it is a bit 'wrong'.

Some people building an event sourced approach or any eventual consistency approach
or even just some other microservices approach that uses distributed state 
might find this sort of plumbing useful though I think for some parts of what they might build.

File: pom.xml (hunk 1 above)
Comment (hutchig) [reply pom.xml]: Done

File: pom.xml (hunk 3 above)
Comment (hutchig) [reply pom.xml]: Done.

File: pom.xml (hunk 5 above)
Comment (hutchig) [reply pom.xml]: Removed

File: pom.xml (hunk 4 above)
Comment (hutchig) [reply pom.xml]: Done

File: pom.xml (hunk 7 above)
Comment (hutchig) [reply pom.xml]: Done

File: pom.xml [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+    </dependencies>
+  </dependencyManagement>
+
+  <dependencies>
+  </dependencies>
+
+  <build>
+    <defaultGoal>clean install</defaultGoal>
+
+    <pluginManagement>
```
Comment (kenfinnigan) [inline pom.xml]: I think the `<pluginManagement>` section can be removed as it's present in the parent

File: pom.xml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+          <artifactId>maven-jar-plugin</artifactId>
+        </plugin>
+      </plugins>
+    </pluginManagement>
+
+    <plugins>
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-javadoc-plugin</artifactId>
+        <executions>
```
Comment (kenfinnigan) [inline pom.xml]: I don't think the config is needed as that should be coming from the parent as well. Just defining the plugin is present for the module should be enough

File: pom.xml [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.smallrye</groupId>
+    <artifactId>smallrye-parent</artifactId>
+    <version>6</version>
+  </parent>
+
+  <groupId>io.smallrye.event-sourcing</groupId>
```
Comment (kenfinnigan) [inline pom.xml]: GroupId should be `io.smallrye` or `io.smallrye.reactive`. They're the only two we have at present

File: pom.xml [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+  <artifactId>smallrye-event-sourcing</artifactId>
+  <version>0.0.10-SNAPSHOT</version>
```
Comment (kenfinnigan) [inline pom.xml]: Change to 1.0.0-SNAPSHOT

File: README.asciidoc [hunk 12]
```diff
@@ -0,0 +1,401 @@
+//
```
Comment (kenfinnigan) [inline README.asciidoc]: This wouldn't be copyright of Eclipse Foundation.

Can remove the entire copyright and notice section

**Review by kenfinnigan — CHANGES_REQUESTED** [review]
Thanks again for making all the changes @hutchig, appreciated

File: pom.xml (hunk 9 above)
Comment (hutchig) [reply pom.xml]: Think I copied this from smallrye reactive messaging but happy to change. Now I have a better sense of the sort of style you want I will tighten up future PRs. This one started out as a  'cp -R' from the source and that was clearly much too informal and the wrong approach for this repo. I appreciate your review patience.

File: pom.xml (hunk 10 above)
Comment (hutchig) [reply pom.xml]: removed, will come from parent.

File: pom.xml (hunk 8 above)
Comment (hutchig) [reply pom.xml]: Done.

File: README.asciidoc (hunk 12 above)
Comment (hutchig) [reply README.asciidoc]: Done

**Comment by hutchig** [comment]
Apologies in advance but I won't typically be able to reply to any comments/questions in this repo between 9 and 5, Monday to Friday at the moment. I am not ignoring you if I don't :-).

File: pom.xml [hunk 13]
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+
+  <parent>
+    <groupId>io.smallrye</groupId>
+    <artifactId>smallrye-parent</artifactId>
+    <version>8-SNAPSHOT</version>
```
Comment (kenfinnigan) [inline pom.xml]: I think this should be the released version, 7?

**Comment by kenfinnigan** [comment]
Thanks for your patience @hutchig

File: pom.xml (hunk 13 above)
Comment (hutchig) [reply pom.xml]: Done.

**Review by kenfinnigan — APPROVED** [review]

**Pull request merged**